vlm/src/main/scala/geotrellis/contrib/vlm/MosaicRasterSource.scala

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2018 Azavea
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package geotrellis.contrib.vlm
+
+import cats.Semigroup
+import cats.implicits._
+import cats.data.NonEmptyList
+import geotrellis.vector._
+import geotrellis.raster._
+import geotrellis.raster.merge.Implicits._
+import geotrellis.raster.resample._
+import geotrellis.raster.reproject.Reproject
+import geotrellis.proj4.{CRS, WebMercator}
+import geotrellis.raster.io.geotiff.{AutoHigherResolution, OverviewStrategy}
+import geotrellis.raster.render._
+import geotrellis.spark.tiling.LayoutDefinition
+import geotrellis.util.GetComponent
+
+/**
+  * Single threaded instance of a reader for reading windows out of collections
+  * of rasters
+  *
+  */
+trait MosaicRasterSource extends RasterSource {
+  // Orphan instance for semigroups for rasters, so we can combine
+  // Option[Raster[_]]s later
+  implicit val rasterSemigroup: Semigroup[Raster[MultibandTile]] =
+    new Semigroup[Raster[MultibandTile]] {
+      def combine(l: Raster[MultibandTile], r: Raster[MultibandTile]) = {
+        val targetRE = RasterExtent(
+          l.rasterExtent.extent combine r.rasterExtent.extent,
+          l.rasterExtent.cellSize)
+        val result = l.resample(targetRE) merge r.resample(targetRE)
+        result
+      }
+    }
+
+  /**
+    * The underlying [[RasterSource]]s that you'll use for data access
+    */
+  val sources: NonEmptyList[RasterSource]
+
+  /**
+    * The commonCrs to reproject all [[RasterSource]]s to anytime we need information about their data
+    *
+    * Since MosaicRasterSources represent collections of [[RasterSource]]s, we don't know in advance
+    * whether they'll have the same CRS. commonCrs allows specifying the CRS on read instead of
+    * having to make sure at compile time that you're threading CRSes through everywhere correctly.
+    */
+  val commonCrs: CRS
+
+  /**
+    * Uri is required only for compatibility with RasterSource.
+    *
+    * It doesn't make sense to access "the" URI for a collection, so this throws an exception.
+    */
+  def uri: String = throw new NotImplementedError(
```

Comment:
I'm wondering if this shouldn't be reflected in changes all the way up top - does that look like a heavy lift?

Reply:
I don't think it makes sense to remove all the way up -- for non-plural `RasterSource`s, requiring a uri is pretty reasonable. Maybe this is a case where having an extra trait for singularity will save us a big headache down the road but I'm not sure.

Reply:
This is probably going to be resolved by expanding the trait hierarchy based on following logic:

- Having URI for RasterSource is useful and reasonable when it's backed by an identifiable resource.
- Having MosaicRasterSource is useful instance of RasterSource interface
- MosaicRasterSource can not be well described by single or even a list of URIs.

=> There is shared interface and there are some traits like `DatasetRasterSource` and `MosaicRasterSource`